File: Python/13-Heap(PriorityQueue)/Algorithms/1-MinHeap-MaxHeap/MinHeap-MaxHeap.py
import heapq  # Python's built-in heap implementation
import time

class Heap:
    """
    This class provides a basic implementation of a Min Heap and Max Heap.
    It includes methods for insertion, extraction, and heapifying a list.
    """

    def __init__(self, heap_type="min"):
        """
        Initializes the Heap object.
        Args:
            heap_type (str):  "min" for Min Heap, "max" for Max Heap.  Defaults to "min".
        """
        if heap_type not in ("min", "max"):
            raise ValueError("heap_type must be 'min' or 'max'")
        self.heap_type = heap_type
        self.data = []  # List to store the heap elements
        self.size = 0  # Current number of elements in the heap

    def __len__(self):
        """
        Returns the number of elements in the heap.  (Allows using len(my_heap))
        """
        return self.size

    def _compare(self, a, b):
        """
        Compares two elements based on the heap type.
        Args:
            a: The first element.
            b: The second element.
        Returns:
            bool: True if a should be considered "smaller" (for min heap)
                  or "larger" (for max heap) than b, False otherwise.
        """
        if self.heap_type == "min":
            return a < b
        else:  # max heap
            return a > b

    def _swap(self, i, j):
        """
        Swaps the elements at indices i and j in the heap's data list.
        Args:
            i: The index of the first element.
            j: The index of the second element.
        """
        self.data[i], self.data[j] = self.data[j], self.data[i]

    def _parent(self, index):
        """
        Calculates the index of the parent node.
        Args:
            index: The index of the child node.
        Returns:
            int: The index of the parent node.
        """
        return (index - 1) // 2

    def _left_child(self, index):
        """
        Calculates the index of the left child node.
        Args:
            index: The index of the parent node.
        Returns:
            int: The index of the left child node.
        """
        return 2 * index + 1

    def _right_child(self, index):
        """
        Calculates the index of the right child node.
        Args:
            index: The index of the parent node.
        Returns:
            int: The index of the right child node.
        """
        return 2 * index + 2

    def _heapify_up(self, index):
        """
        Performs the heapify-up operation, moving a node up the tree
        to its correct position.  Used after insertion.
        Args:
            index: The index of the node to heapify up.
        """
        while index > 0 and self._compare(self.data[index], self.data[self._parent(index)]):
            self._swap(index, self._parent(index))
            index = self._parent(index)

    def _heapify_down(self, index):
        """
        Performs the heapify-down operation, moving a node down the tree
        to its correct position.  Used after extraction (removing the root).
        Args:
            index: The index of the node to heapify down.
        """
        left = self._left_child(index)
        right = self._right_child(index)
        largest = index  # or smallest, depending on heap type

        if left < self.size and self._compare(self.data[left], self.data[largest]):
            largest = left
        if right < self.size and self._compare(self.data[right], self.data[largest]):
            largest = right

        if largest != index:
            self._swap(index, largest)
            self._heapify_down(largest)

    def insert(self, value):
        """
        Inserts a new value into the heap.
        Args:
            value: The value to insert.
        """
        self.data.append(value)
        self.size += 1
        self._heapify_up(self.size - 1)

    def extract(self):
        """
        Extracts (removes and returns) the root element from the heap
        (which is the minimum element in a Min Heap, or the maximum
        element in a Max Heap).
        Returns:
            The root element of the heap, or None if the heap is empty.
        """
        if self.size == 0:
            return None  # Handle empty heap case
        if self.size == 1:
            self.size -= 1
            return self.data.pop()

        root = self.data[0]
        self.data[0] = self.data[self.size - 1]  # Move last element to root
        self.data.pop()
        self.size -= 1
        self._heapify_down(0)  # Restore heap property
        return root

    def peek(self):
        """
        Returns the root element of the heap without removing it.
        Returns:
            The root element, or None if the heap is empty.
        """
        if self.size == 0:
            return None
        return self.data[0]

    def display(self):
        """
        Prints the elements of the heap in their current order.  This is
        for debugging and visualization, and does *not* represent the
        sorted order.
        """
        print(f"{self.heap_type.title()} Heap: {self.data[:self.size]}")

    def heapify(self, arr):
        """
        Initializes the heap from an existing list (array).  This is a
        more efficient way to build a heap than inserting elements one by one.
        Args:
           arr:  The list to heapify.
        """
        self.data = arr[:]  # Important: create a *copy* of the list.
        self.size = len(arr)
        # Start from the last non-leaf node and heapify downwards.
        for i in range(self.size // 2 - 1, -1, -1):
            self._heapify_down(i)
        # added return self
        return self

def heap_sort(arr, order="asc"):
    """
    Sorts a list using the Heap Sort algorithm.
    Args:
        arr (list): The list to be sorted.
        order (str): "asc" for ascending, "desc" for descending.
    Returns:
        list: The sorted list.
    """
    if order not in ("asc", "desc"):
        raise ValueError("order must be 'asc' or 'desc'")

    n = len(arr)
    if order == "asc":
        heap = Heap("min")  # Use Min Heap for ascending order
    else:
        heap = Heap("max")  # Use Max Heap for descending order

    heap.heapify(arr)  # Efficiently build the heap
    sorted_arr = []
    for _ in range(n):
        sorted_arr.append(heap.extract())  # Extract elements in sorted order
    return sorted_arr

def main():
    """
    Main function to demonstrate the usage of the Heap class and heap sort.
    """
    print("--- Heap Implementation Demonstration ---")

    # 1. Using the Heap class (Min Heap)
    print("\n1. Using the Heap class (Min Heap):")
    min_heap = Heap("min")
    min_heap.insert(5)
    min_heap.insert(3)
    min_heap.insert(7)
    min_heap.insert(1)
    min_heap.insert(9)
    min_heap.display()  # Output: Min Heap: [1, 3, 5, 7, 9]
    print("Peek:", min_heap.peek())  # Output: Peek: 1
    print("Extract:", min_heap.extract())  # Output: Extract: 1
    min_heap.display()  # Output: Min Heap: [3, 5, 7, 9]

    # 2. Using the Heap class (Max Heap)
    print("\n2. Using the Heap class (Max Heap):")
    max_heap = Heap("max")
    max_heap.insert(5)
    max_heap.insert(3)
    max_heap.insert(7)
    max_heap.insert(1)
    max_heap.insert(9)
    max_heap.display()  # Output: Max Heap: [9, 7, 5, 1, 3]
    print("Peek:", max_heap.peek())  # Output: Peek: 9
    print("Extract:", max_heap.extract())  # Output: Extract: 9
    max_heap.display()  # Output: Max Heap: [7, 5, 3, 1]

    # 3. Heapify an existing list (Min Heap)
    print("\n3. Heapify an existing list (Min Heap):")
    my_list = [5, 3, 7, 1, 9]
    min_heap2 = Heap("min")
    min_heap2.heapify(my_list)
    min_heap2.display()  # Output: Min Heap: [1, 3, 5, 7, 9]

    # 4. Heap Sort (Ascending Order)
    print("\n4. Heap Sort (Ascending Order):")
    unsorted_list = [5, 3, 7, 1, 9, 2, 8, 4, 6, 0]
    sorted_list_asc = heap_sort(unsorted_list, "asc")
    print("Unsorted List:", unsorted_list)
    print("Sorted List (Ascending):", sorted_list_asc)  # Output: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

    # 5. Heap Sort (Descending Order)
    print("\n5. Heap Sort (Descending Order):")
    unsorted_list = [5, 3, 7, 1, 9, 2, 8, 4, 6, 0]
    sorted_list_desc = heap_sort(unsorted_list, "desc")
    print("Unsorted List:", unsorted_list)
    print("Sorted List (Descending):", sorted_list_desc)  # Output: [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

    # 6. Using Python's built-in heapq (Min Heap) - Demonstration
    print("\n6. Using Python's built-in heapq (Min Heap):")
    my_list = [5, 3, 7, 1, 9]
    print("Original list:", my_list)
    heapq.heapify(my_list)  # In-place heapify
    print("Heapified list (using heapq):", my_list)  # Output: [1, 3, 5, 7, 9]

    # Push and pop elements
    heapq.heappush(my_list, 2)
    print("List after pushing 2:", my_list) # Output: [1, 2, 5, 7, 9, 3]
    print("Popped element:", heapq.heappop(my_list))  # Output: 1
    print("List after popping:", my_list) #  Output: [2, 3, 5, 7, 9]

if __name__ == "__main__":
    main()

File: Java/13-Heap(PriorityQueue)/Algorithms/1-MinHeap-MaxHeap/MinHeap-MaxHeap.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Comparator;

public class HeapImplementation {

    // Approach 1: Using Java's built-in PriorityQueue (Min Heap by default)
    public static class Approach1 {
        public static void demonstrate() {
            System.out.println("\nApproach 1: Using Java's built-in PriorityQueue (Min Heap)");
            // Creates a min heap
            PriorityQueue<Integer> minHeap = new PriorityQueue<>();

            // Insert elements
            minHeap.add(5);
            minHeap.add(10);
            minHeap.add(3);
            minHeap.add(8);
            minHeap.add(1);

            // Peek and print the minimum element
            System.out.println("Minimum element: " + minHeap.peek()); // Output: 1

            // Poll (remove) and print elements in sorted order (ascending for min heap)
            System.out.print("Elements in ascending order: ");
            while (!minHeap.isEmpty()) {
                System.out.print(minHeap.poll() + " "); // Output: 1 3 5 8 10
            }
            System.out.println();
        }
    }

    // Approach 2: Using Java's built-in PriorityQueue (Max Heap using Comparator)
    public static class Approach2 {
        public static void demonstrate() {
            System.out.println("\nApproach 2: Using Java's built-in PriorityQueue (Max Heap)");
            // Creates a max heap using a custom comparator
            PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());

            // Insert elements
            maxHeap.add(5);
            maxHeap.add(10);
            maxHeap.add(3);
            maxHeap.add(8);
            maxHeap.add(1);

            // Peek and print the maximum element
            System.out.println("Maximum element: " + maxHeap.peek()); // Output: 10

            // Poll (remove) and print elements in sorted order (descending for max heap)
            System.out.print("Elements in descending order: ");
            while (!maxHeap.isEmpty()) {
                System.out.print(maxHeap.poll() + " "); // Output: 10 8 5 3 1
            }
            System.out.println();
        }
    }

    // Approach 3: Implementing a Min Heap from scratch using an ArrayList
    public static class Approach3 {

        public static class MinHeap {
            private List<Integer> heap;

            public MinHeap() {
                this.heap = new ArrayList<>();
            }

            // Helper method to get the parent index
            private int getParent(int index) {
                return (index - 1) / 2;
            }

            // Helper method to get the left child index
            private int getLeftChild(int index) {
                return 2 * index + 1;
            }

            // Helper method to get the right child index
            private int getRightChild(int index) {
                return 2 * index + 2;
            }

            // Helper method to swap two elements
            private void swap(int i, int j) {
                int temp = heap.get(i);
                heap.set(i, heap.get(j));
                heap.set(j, temp);
            }

            // Method to insert a new element into the heap
            public void insert(int value) {
                heap.add(value);
                int index = heap.size() - 1;
                // Maintain heap property (min heap)
                while (index > 0 && heap.get(index) < heap.get(getParent(index))) {
                    swap(index, getParent(index));
                    index = getParent(index);
                }
            }

            // Method to remove the minimum element from the heap
            public Integer remove() {
                if (heap.isEmpty()) {
                    return null; // Or throw an exception: throw new NoSuchElementException("Heap is empty");
                }
                if (heap.size() == 1) {
                    return heap.remove(0);
                }

                int root = heap.get(0);
                heap.set(0, heap.get(heap.size() - 1)); // Move the last element to the root
                heap.remove(heap.size() - 1);         // Remove the last element
                minHeapify(0);                   // Restore the min heap property
                return root;
            }

            // Method to maintain the min heap property (downward)
            private void minHeapify(int index) {
                int left = getLeftChild(index);
                int right = getRightChild(index);
                int smallest = index;

                if (left < heap.size() && heap.get(left) < heap.get(smallest)) {
                    smallest = left;
                }
                if (right < heap.size() && heap.get(right) < heap.get(smallest)) {
                    smallest = right;
                }

                if (smallest != index) {
                    swap(index, smallest);
                    minHeapify(smallest);
                }
            }

            // Method to peek the minimum element
            public Integer peek() {
                if (heap.isEmpty()) {
                    return null;
                }
                return heap.get(0);
            }

            public boolean isEmpty() {
                return heap.isEmpty();
            }
        }

        public static void demonstrate() {
            System.out.println("\nApproach 3: Implementing a Min Heap from scratch using an ArrayList");
            MinHeap minHeap = new MinHeap();

            minHeap.insert(5);
            minHeap.insert(10);
            minHeap.insert(3);
            minHeap.insert(8);
            minHeap.insert(1);

            System.out.println("Minimum element: " + minHeap.peek());

            System.out.print("Elements in ascending order: ");
            while (!minHeap.isEmpty()) {
                System.out.print(minHeap.remove() + " ");
            }
            System.out.println();
        }
    }

    // Approach 4: Implementing a Max Heap from scratch using an ArrayList
    public static class Approach4 {

        public static class MaxHeap {
            private List<Integer> heap;

            public MaxHeap() {
                this.heap = new ArrayList<>();
            }

            private int getParent(int index) {
                return (index - 1) / 2;
            }

            private int getLeftChild(int index) {
                return 2 * index + 1;
            }

            private int getRightChild(int index) {
                return 2 * index + 2;
            }

            private void swap(int i, int j) {
                int temp = heap.get(i);
                heap.set(i, heap.get(j));
                heap.set(j, temp);
            }

            public void insert(int value) {
                heap.add(value);
                int index = heap.size() - 1;
                while (index > 0 && heap.get(index) > heap.get(getParent(index))) {
                    swap(index, getParent(index));
                    index = getParent(index);
                }
            }

            public Integer remove() {
                if (heap.isEmpty()) {
                    return null;
                }
                if (heap.size() == 1) {
                    return heap.remove(0);
                }
                int root = heap.get(0);
                heap.set(0, heap.get(heap.size() - 1));
                heap.remove(heap.size() - 1);
                maxHeapify(0);
                return root;
            }

            private void maxHeapify(int index) {
                int left = getLeftChild(index);
                int right = getRightChild(index);
                int largest = index;

                if (left < heap.size() && heap.get(left) > heap.get(largest)) {
                    largest = left;
                }
                if (right < heap.size() && heap.get(right) > heap.get(largest)) {
                    largest = right;
                }

                if (largest != index) {
                    swap(index, largest);
                    maxHeapify(largest);
                }
            }
            public Integer peek() {
                if (heap.isEmpty()) {
                    return null;
                }
                return heap.get(0);
            }

            public boolean isEmpty() {
                return heap.isEmpty();
            }
        }

        public static void demonstrate() {
            System.out.println("\nApproach 4: Implementing a Max Heap from scratch using an ArrayList");
            MaxHeap maxHeap = new MaxHeap();

            maxHeap.insert(5);
            maxHeap.insert(10);
            maxHeap.insert(3);
            maxHeap.insert(8);
            maxHeap.insert(1);

            System.out.println("Maximum element: " + maxHeap.peek());

            System.out.print("Elements in descending order: ");
            while (!maxHeap.isEmpty()) {
                System.out.print(maxHeap.remove() + " ");
            }
            System.out.println();
        }
    }

    // Approach 5: Using heapify to build a Min Heap from an unsorted array
    public static class Approach5 {
        public static void demonstrate() {
            System.out.println("\nApproach 5: Using heapify to build a Min Heap from an unsorted array");
            int[] arr = {5, 10, 3, 8, 1};
            buildMinHeap(arr);

            System.out.print("Elements after building Min Heap: ");
            for (int value : arr) {
                System.out.print(value + " ");
            }
            System.out.println();

            System.out.print("Elements in ascending order using heap sort: ");
            int[] sortedArray = heapSort(arr.clone()); // Clone to avoid modifying the original
            for(int value : sortedArray){
                System.out.print(value + " ");
            }
            System.out.println();
        }

        // Method to build a Min Heap from an unsorted array
        public static void buildMinHeap(int[] arr) {
            int n = arr.length;
            // Start from the last non-leaf node and heapify all preceding nodes
            for (int i = n / 2 - 1; i >= 0; i--) {
                minHeapify(arr, n, i);
            }
        }

        private static void minHeapify(int[] arr, int n, int i) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            int smallest = i;

            if (left < n && arr[left] < arr[smallest]) {
                smallest = left;
            }
            if (right < n && arr[right] < arr[smallest]) {
                smallest = right;
            }

            if (smallest != i) {
                swap(arr, i, smallest);
                minHeapify(arr, n, smallest);
            }
        }

        private static void swap(int[] arr, int i, int j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }

        public static int[] heapSort(int[] arr) {
            int n = arr.length;

            // Build min heap (in-place)
            buildMinHeap(arr);

            // Create a new array for the sorted result
            int[] sortedArray = new int[n];
            for (int i = 0; i < n; i++) {
                sortedArray[i] = removeMin(arr, n - i); // Remove min from heap and store in sortedArray
            }
            return sortedArray;
        }

        // Function to remove the minimum element from the heap (used in heap sort)
        private static int removeMin(int[] arr, int n) {
            int min = arr[0];
            arr[0] = arr[n - 1]; // Replace root with the last element in the (reduced) heap
            minHeapify(arr, n - 1, 0); // Heapify the root
            return min;
        }
    }

    public static void main(String[] args) {
        Approach1.demonstrate(); // Min Heap using PriorityQueue
        Approach2.demonstrate(); // Max Heap using PriorityQueue
        Approach3.demonstrate(); // Min Heap from scratch
        Approach4.demonstrate(); // Max Heap from scratch
        Approach5.demonstrate(); // Build Min Heap from unsorted array and Heap Sort
    }
}

